### PR TITLE
Fix constant value deactivation

### LINKED ISSUE #1330: Should not deactivate a constant value that has not been activated
https://github.com/inversify/InversifyJS/blob/9f6f43e53752679a610684baceea98533ad65c4b/src/container/container.ts#L640

Change to 
```
private _deactivateIfSingleton(binding: Binding<any>): Promise<void> | void {
        if (!binding.activated) {
            return;
        }
```

### PR DESCRIPTION
**Pull request by tonyhallett**
fix #1330

### CODE REVIEW

File: test/resolution/resolver.test.ts [hunk 1]
```diff
@@ -1358,6 +1358,15 @@ describe("Resolve", () => {
       expect(subject1 === subject2).eql(true);
   });
 
+  it("should not deactivate a non activated constant value", () => {
+    const container = new Container();
+    let deactivated = false;
+    container.bind<string>("ConstantValue").toConstantValue("Constant").onDeactivation(_ => {deactivated = true;})
+    container.unbind("ConstantValue");
+    expect(deactivated).to.equal(false);
+  });
```
Comment (notaphplover) [inline test/resolution/resolver.test.ts]: I'd suggest using sinnon spies next time, but we can afford not to do it now

**Review by notaphplover — APPROVED** [review]
Nice! LGTM

File: src/container/container.ts [hunk 2]
```diff
@@ -637,7 +637,7 @@ class Container implements interfaces.Container {
     }
 
     private _deactivateIfSingleton(binding: Binding<any>): Promise<void> | void {
```
Comment (notaphplover) [inline src/container/container.ts:639]: Maybe this method should be called `_deactivateIfActivated` instead

File: test/resolution/resolver.test.ts (hunk 1 above)
Comment (tonyhallett) [reply test/resolution/resolver.test.ts]: but then would have had got access to the binding.

File: src/container/container.ts (hunk 2 above)
Comment (tonyhallett) [reply src/container/container.ts:639]: It is clear that deactivation applies to singletons only with _deactivateIfSingleton.  I think activated is a bad name choice given anything can be "activated" through onActivation.  

The [v6-binding-abstractions](https://github.com/inversify/InversifyJS/pull/1336) change I made hopefully makes it clearer.

```
    private _deactivateIfSingleton(binding: interfaces.Binding<any>): Promise<void> | void {
        const scope = binding.scope;
        if(scope.type === ConfigurableBindingScopeEnum.Singleton){
            const cached = scope.resolved; // discriminated union to access directly
            if (cached === undefined) {
                return;
            }

            if (isPromise(cached)) {
                return cached.then((resolved: any) => this._deactivate(binding, resolved));
            }

            return this._deactivate(binding, cached);
        }

    }

```

File: test/resolution/resolver.test.ts (hunk 1 above)
Comment (notaphplover) [reply test/resolution/resolver.test.ts]: Sorry I don't understand your point :( (my english is far to be perfect)

Comment (tonyhallett) [reply test/resolution/resolver.test.ts]: > Sorry I don't understand your point :( (my english is far to be perfect)

There was no point to my point !

You are correct, I should have done
```

  it("should not deactivate a non activated constant value", () => {
    const container = new Container();
    const deactivationHandler = sinon.spy();
    container.bind<string>("ConstantValue").toConstantValue("Constant").onDeactivation(deactivationHandler)
    container.unbind("ConstantValue");
    // tslint:disable-next-line: no-unused-expression
    expect(deactivationHandler.notCalled).to.be.true;
  });
```

Will change

Comment (tonyhallett) [reply test/resolution/resolver.test.ts]: Done

**Review by notaphplover — APPROVED** [review]

**Pull request merged**